hogql_parser/parser.cpp

```diff
@@ -0,0 +1,1358 @@
+    }
+    return visit(ctx->selectStmt());
+  }
+
+  VISIT(SelectStmtWithParens) {
+    auto select_stmt_ctx = ctx->selectStmt();
+    if (select_stmt_ctx) {
+      return visit(select_stmt_ctx);
+    }
+    return visit(ctx->selectUnionStmt());
+  }
+
+  VISIT(SelectUnionStmt) {
+    // Using a vector of PyObjects atypically here, because this is a precursor of flattened_queries
+    vector<PyObject*> select_queries;
+    auto select_stmt_with_parens_ctxs = ctx->selectStmtWithParens();
+    select_queries.reserve(select_stmt_with_parens_ctxs.size());
+    for (auto select_stmt_with_parens_ctx : select_stmt_with_parens_ctxs) {
+      select_queries.push_back(visitAsPyObject(select_stmt_with_parens_ctx));
+    }
+    PyObject* flattened_queries = PyList_New(0);
+    for (auto query : select_queries) {
+      if (is_ast_node_instance(query, "SelectQuery")) {
+        PyList_Append(flattened_queries, query);
+      } else if (is_ast_node_instance(query, "SelectUnionQuery")) {
+        // Extend flattened_queries with sub_select_queries
+        PyObject* sub_select_queries = PyObject_GetAttrString(query, "select_queries");
+        X_PyList_Extend(flattened_queries, sub_select_queries);
+        Py_DECREF(sub_select_queries);
+      } else {
+        Py_DECREF(flattened_queries);
+        throw HogQLParsingException("Unexpected query node type: " + string(Py_TYPE(query)->tp_name));
+      }
+    }
+    if (PyList_Size(flattened_queries) == 1) {
+      PyObject* query = PyList_GET_ITEM(flattened_queries, 0);
+      Py_INCREF(query);
+      Py_DECREF(flattened_queries);
+      return query;
+    }
+    return build_ast_node("SelectUnionQuery", "{s:N}", "select_queries", flattened_queries);
+  }
+
+  VISIT(SelectStmt) {
+    PyObject* select_query = build_ast_node(
+        "SelectQuery", "{s:N,s:N,s:N,s:N,s:N,s:N,s:N,s:N,s:N}", "ctes", visitAsPyObjectOrNone(ctx->withClause()),
+        "select", visitAsPyObjectOrEmptyList(ctx->columnExprList()), "distinct",
+        Py_NewRef(ctx->DISTINCT() ? Py_True : Py_None), "select_from", visitAsPyObjectOrNone(ctx->fromClause()),
+        "where", visitAsPyObjectOrNone(ctx->whereClause()), "prewhere", visitAsPyObjectOrNone(ctx->prewhereClause()),
+        "having", visitAsPyObjectOrNone(ctx->havingClause()), "group_by", visitAsPyObjectOrNone(ctx->groupByClause()),
+        "order_by", visitAsPyObjectOrNone(ctx->orderByClause())
+    );
+
+    auto window_clause_ctx = ctx->windowClause();
+    if (window_clause_ctx) {
+      auto window_expr_ctxs = window_clause_ctx->windowExpr();
+      auto identifier_ctxs = window_clause_ctx->identifier();
+      PyObject* window_exprs = PyDict_New();
+      PyObject_SetAttrString(select_query, "window_exprs", window_exprs);
+      for (size_t i = 0; i < window_expr_ctxs.size(); i++) {
+        PyDict_SetItemString(
+            window_exprs, visitAsString(identifier_ctxs[i]).c_str(), visitAsPyObject(window_expr_ctxs[i])
+        );
+      }
```

Comment:
The grammar dictates that we have the same count of window exprs and identifiers, but without double checking that it's the case, I'm a bit nervous about reading out of bounds memory with `identifier_ctxs[i]`. Probably fine though 😅

Reply:
Currently a parse tree where this doesn't hold is impossible. But good to be defensive, I added a check to ensure this also holds in the future

Reply:
Also added a similar check in `RatioExpr`